src/RGBEncoder.cpp

```diff
@@ -0,0 +1,141 @@
+/*
+ * This file is a part of RGBEncoder library.
+ *
+ * Product page: https://amperka.ru/product/rgbencoder
+ * © Amperka LLC (https://amperka.com, dev@amperka.com)
+ *
+ * Author: Yury Botov <by@amperka.com>
+ * License: GPLv3, all text here must be included in any redistribution.
+ */
+
+#include "RGBEncoder.h"
+
+void buttonDumb(bool state) { (void)state; }
+void encoderDumb(int32_t value) { (void)value; }
+void stepsDumb() { }
+
+RGBEncoder::RGBEncoder(uint8_t i2cAddress) {
+    _i2cAddress = i2cAddress;
+    _buttonState = false;
+    _encoderLowerBound = 0;
+    _encoderUpperBound = 255;
+    _encoderCurrentValue = 0;
+    _encoderStep = 1;
+    _encoderMode = RGBENC_CLAMP;
+    _buttonHandler = buttonDumb;
+    _encoderHandler = encoderDumb;
+    _stepLeftHandler = stepsDumb;
+    _stepRightHandler = stepsDumb;
+}
+
+void RGBEncoder::begin() { begin(&Wire); }
+
+void RGBEncoder::begin(TwoWire* wire) {
+    wire->begin();
+    GpioExpander::begin(wire);
+
+    GpioExpander::pinMode(_redLedPin, OUTPUT);
+    GpioExpander::analogWrite(_redLedPin, 0);
+    GpioExpander::pinMode(_greenLedPin, OUTPUT);
+    GpioExpander::analogWrite(_greenLedPin, 0);
+    GpioExpander::pinMode(_blueLedPin, OUTPUT);
+    GpioExpander::analogWrite(_blueLedPin, 0);
+
+    GpioExpander::pinMode(_encoderAPin, INPUT);
+    GpioExpander::pinMode(_encoderBPin, INPUT);
+    GpioExpander::setEncoderPins(0, _encoderAPin, _encoderBPin);
+
+    GpioExpander::pinMode(_buttonPin, INPUT);
+    int16_t state = GpioExpander::digitalRead(_buttonPin);
+    if (state == 0)
+        _buttonState = false;
+    else
+        _buttonState = true;
+}
+
+void RGBEncoder::changeAddr(uint8_t newAddress) {
+    GpioExpander::changeAddr(newAddress);
+}
+
+void RGBEncoder::saveAddr() {
+    GpioExpander::saveAddr();
+}
+
+void RGBEncoder::setValue(int32_t value) { _encoderCurrentValue = value; }
+
+void RGBEncoder::setRange(int32_t min, int32_t max) {
+    _encoderLowerBound = min;
+    _encoderUpperBound = max;
+}
+
+void RGBEncoder::setStep(int32_t step) { _encoderStep = step; }
+
+void RGBEncoder::setMode(enum EncoderMode mode) { _encoderMode = mode; }
+
+void RGBEncoder::update() {
+    // button part
+    int16_t state = GpioExpander::digitalRead(_buttonPin);
+    if (state == 0 && _buttonState == true) {
```

Comment:
```suggestion
    if (state == LOW && _buttonState) {
```

Reply:
need to test it. int16_t GpioExpander::digitalRead(_buttonPin) is very strange method

Reply:
Done